### ISSUE TITLE
toke.c:3021 format string warning under GCC 5.2

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14897]
Migrated from [rt.perl.org#126036](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126036) (status was 'resolved')

Searchable as RT126036$

**Comment by p5pRT** [comment]
### From @dcollinsn

Greetings\,

At toke\.c&#8203;:3021\, we want to print hexadecimal representations of IVs \(range\_min and range\_max\) as part of an error message&#8203;:

&nbsp;    else if \(convert\_unicode\) \{
&nbsp;        /\* diag\_listed\_as&#8203;: Invalid range "%s" in transliteration operator \*/
&nbsp;        Perl\_croak\(aTHX\_
&nbsp;   "Invalid range \\"\\\\N\{U\+%04X\}\-\\\\N\{U\+%04X\}\\""
&nbsp;               " in transliteration operator"\,
&nbsp;   range\_min\, range\_max\);
&nbsp;    \}
&nbsp;    else \{
&nbsp;        /\* diag\_listed\_as&#8203;: Invalid range "%s" in transliteration operator \*/
&nbsp;        Perl\_croak\(aTHX\_
&nbsp;   "Invalid range \\"\\\\x\{%04X\}\-\\\\x\{%04X\}\\""
&nbsp;               " in transliteration operator"\,
&nbsp;   range\_min\, range\_max\);
&nbsp;    \}

However on some platforms\, including at a minimum my \-Duse64bitint build on x86\, IV is a long long\. Under GCC 5\.2\, this throws the following warning&#8203;:

toke\.c&#8203;: In function ‘S\_scan\_const’&#8203;:
toke\.c&#8203;:3021&#8203;:11&#8203;: warning&#8203;: format ‘%X’ expects argument of type ‘unsigned int’\, but argument 2 has type ‘IV \{aka long long int\}’ \[\-Wformat=\]
&nbsp;           "Invalid range \\"\\\\N\{U\+%04X\}\-\\\\N\{U\+%04X\}\\""
&nbsp;           ^
toke\.c&#8203;:3021&#8203;:11&#8203;: warning&#8203;: format ‘%X’ expects argument of type ‘unsigned int’\, but argument 3 has type ‘IV \{aka long long int\}’ \[\-Wformat=\]
toke\.c&#8203;:3028&#8203;:11&#8203;: warning&#8203;: format ‘%X’ expects argument of type ‘unsigned int’\, but argument 2 has type ‘IV \{aka long long int\}’ \[\-Wformat=\]
&nbsp;           "Invalid range \\"\\\\x\{%04X\}\-\\\\x\{%04X\}\\""
&nbsp;           ^
toke\.c&#8203;:3028&#8203;:11&#8203;: warning&#8203;: format ‘%X’ expects argument of type ‘unsigned int’\, but argument 3 has type ‘IV \{aka long long int\}’ \[\-Wformat=\]

To fix this warning\, I have modified toke\.c and config\.h\. I know config\.h is autogenerated\, however I don't know how to influence the content of Configure and config\_h\.SH\, so I have no idea how to make these changes permanent\. The necessary line in config\.h is&#8203;:

\#define IVXf            "LX"            /\*\*/

And the diff&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/toke.c b/toke.c
index 27e6650..d3fdf4b 100644
--- a/toke.c
+++ b/toke.c
@@ -3018,14 +3018,14 @@ S_scan_const(pTHX_ char *start)
                     else if (convert_unicode) {
                         /* diag_listed_as: Invalid range "%s" in transliteration operator */
                         Perl_croak(aTHX_
-			       "Invalid range \"\\N{U+%04X}-\\N{U+%04X}\""
+			       "Invalid range \"\\N{U+%04"IVXf"}-\\N{U+%04"IVXf"}\""
                                " in transliteration operator",
 			       range_min, range_max);
                     }
                     else {
                         /* diag_listed_as: Invalid range "%s" in transliteration operator */
                         Perl_croak(aTHX_
-			       "Invalid range \"\\x{%04X}-\\x{%04X}\""
+			       "Invalid range \"\\x{%04"IVXf"}-\\x{%04"IVXf"}\""
                                " in transliteration operator",
 			       range_min, range_max);
                     }


NB: Technically I think that these long longs should have a format argument of llX and not LX, as L is supposed to be a long double. However, Configure uses L* for IVs, UVs, and NVs on my platform, and doesn't warn when passing range_min (a long long int) to LX. So, whatever the hell you actually do to make Configure generate $ivXformat, it need only be $ivdformat =~ tr/d/X/, no new logic should be required.
```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

On 09/10/2015 08&#8203;:15 PM\, Dan Collins \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Dan Collins
> \# Please include the string&#8203;:  \[perl \#126036\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126036 >
>
>
> Greetings\,
>
> At toke\.c&#8203;:3021\, we want to print hexadecimal representations of IVs \(range\_min and range\_max\) as part of an error message&#8203;:
>
>      else if \(convert\_unicode\) \{
>          /\* diag\_listed\_as&#8203;: Invalid range "%s" in transliteration operator \*/
>          Perl\_croak\(aTHX\_
>     "Invalid range \\"\\\\N\{U\+%04X\}\-\\\\N\{U\+%04X\}\\""
>                 " in transliteration operator"\,
>     range\_min\, range\_max\);
>      \}
>      else \{
>          /\* diag\_listed\_as&#8203;: Invalid range "%s" in transliteration operator \*/
>          Perl\_croak\(aTHX\_
>     "Invalid range \\"\\\\x\{%04X\}\-\\\\x\{%04X\}\\""
>                 " in transliteration operator"\,
>     range\_min\, range\_max\);
>      \}
>
> However on some platforms\, including at a minimum my \-Duse64bitint build on x86\, IV is a long long\. Under GCC 5\.2\, this throws the following warning&#8203;:
>
> toke\.c&#8203;: In function ‘S\_scan\_const’&#8203;:
> toke\.c&#8203;:3021&#8203;:11&#8203;: warning&#8203;: format ‘%X’ expects argument of type ‘unsigned int’\, but argument 2 has type ‘IV \{aka long long int\}’ \[\-Wformat=\]
>             "Invalid range \\"\\\\N\{U\+%04X\}\-\\\\N\{U\+%04X\}\\""
>             ^
> toke\.c&#8203;:3021&#8203;:11&#8203;: warning&#8203;: format ‘%X’ expects argument of type ‘unsigned int’\, but argument 3 has type ‘IV \{aka long long int\}’ \[\-Wformat=\]
> toke\.c&#8203;:3028&#8203;:11&#8203;: warning&#8203;: format ‘%X’ expects argument of type ‘unsigned int’\, but argument 2 has type ‘IV \{aka long long int\}’ \[\-Wformat=\]
>             "Invalid range \\"\\\\x\{%04X\}\-\\\\x\{%04X\}\\""
>             ^
> toke\.c&#8203;:3028&#8203;:11&#8203;: warning&#8203;: format ‘%X’ expects argument of type ‘unsigned int’\, but argument 3 has type ‘IV \{aka long long int\}’ \[\-Wformat=\]
>
> To fix this warning\, I have modified toke\.c and config\.h\. I know config\.h is autogenerated\, however I don't know how to influence the content of Configure and config\_h\.SH\, so I have no idea how to make these changes permanent\. The necessary line in config\.h is&#8203;:
>
> \#define IVXf            "LX"            /\*\*/
>
> And the diff&#8203;:
>
> diff \-\-git a/toke\.c b/toke\.c
> index 27e6650\.\.d3fdf4b 100644
> \-\-\- a/toke\.c
> \+\+\+ b/toke\.c
> @&#8203;@&#8203; \-3018\,14 \+3018\,14 @&#8203;@&#8203; S\_scan\_const\(pTHX\_ char \*start\)
>                       else if \(convert\_unicode\) \{
>                           /\* diag\_listed\_as&#8203;: Invalid range "%s" in transliteration operator \*/
>                           Perl\_croak\(aTHX\_
> \-                   "Invalid range \\"\\\\N\{U\+%04X\}\-\\\\N\{U\+%04X\}\\""
> \+                   "Invalid range \\"\\\\N\{U\+%04"IVXf"\}\-\\\\N\{U\+%04"IVXf"\}\\""
>                                  " in transliteration operator"\,
>                      range\_min\, range\_max\);
>                       \}
>                       else \{
>                           /\* diag\_listed\_as&#8203;: Invalid range "%s" in transliteration operator \*/
>                           Perl\_croak\(aTHX\_
> \-                   "Invalid range \\"\\\\x\{%04X\}\-\\\\x\{%04X\}\\""
> \+                   "Invalid range \\"\\\\x\{%04"IVXf"\}\-\\\\x\{%04"IVXf"\}\\""
>                                  " in transliteration operator"\,
>                      range\_min\, range\_max\);
>                       \}
>
>
> NB&#8203;: Technically I think that these long longs should have a format argument of llX and not LX\, as L is supposed to be a long double\. However\, Configure uses L\* for IVs\, UVs\, and NVs on my platform\, and doesn't warn when passing range\_min \(a long long int\) to LX\. So\, whatever the hell you actually do to make Configure generate $ivXformat\, it need only be $ivdformat =~ tr/d/X/\, no new logic should be required\.
>
>

Even though these are declared IV\, they can never be negative\, so the 
existing UVXf format works\.  I have pushed a smoke branch with that fix as
http&#8203;://perl5\.git\.perl\.org/perl\.git/shortlog/refs/heads/smoke\-me/khw\-regcomp

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for reporting this\.  Fixed in 1f9a1c0bdb88fe8f4b38dadcebb4988215612692
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'
